lib/TaskProcessing/SummaryProvider.php

```diff
@@ -114,21 +119,45 @@ public function process(?string $userId, array $input, callable $reportProgress)
 			$model = $this->appConfig->getValueString(Application::APP_ID, 'default_completion_model_id', Application::DEFAULT_MODEL_ID) ?: Application::DEFAULT_MODEL_ID;
 		}
 
-		try {
-			if ($this->openAiAPIService->isUsingOpenAi() || $this->openAiSettingsService->getChatEndpointEnabled()) {
-				$completion = $this->openAiAPIService->createChatCompletion($userId, $model, $prompt, null, null, 1, $maxTokens);
-			} else {
-				$completion = $this->openAiAPIService->createCompletion($userId, $prompt, 1, $model, $maxTokens);
+		while (strlen($prompt) > $maxTokens) {
+			$prompts = [$wrapSummaryPrompt($prompt)];
+			$promptIsChunked = $chunkSize && (strlen($prompt) > $chunkSize);
+			if ($promptIsChunked) {
+				$prompts = explode('<|split|>', chunk_split($prompt, $chunkSize, '<|split|>'));
```

Comment:
Does `chunk_split` split along sentence boundaries? AFAIK it splits cleanly after x characters/bytes. This will break sentences at the boundaries of chunks. Can we implement a proper sentence boundary splitter?

Reply:
Additionally, I think the chunk size should be measured in tokens not in characters, so that admins can easily set the value from a models context size which is usually given in tokens.

Reply:
(One word is approximately 1.5 tokens on average, I believe.)

Reply:
> Can we implement a proper sentence boundary splitter?

:100:

exploding on `newlines -> fullstops and other delimiters -> words`
would be efficient in packing the chunks for a given chunk size, see what fits best, maybe the delimiter split will be troubling in blocks of code.

tokens can vary based on languages, see the token ratio section here: https://gptforwork.com/guides/openai-gpt3-tokens. Maybe we can relax it a little more like 2 to 2.5 tokens, considering multilingual text and occasional code blocks?

-- or, we can call a local GPT2 fast tokenizer to get the exact tokens, like we do in context chat backend (implicitly).

Reply:
2.5 sounds good, we can overestimate, but we shouldn't underestimate

Reply:
Why not use a similar algorithm that langchain uses since we use their text splitter for LLM2?

https://python.langchain.com/api_reference/_modules/langchain_text_splitters/character.html#CharacterTextSplitter.split_text

Reply:
Loop is now rewritten to split on `newlines -> sentences -> words -> characters`. The only thing left that's bugging me at the moment is that when I test the assistant with longer text, the splitting doesn't actually work. I tested the function in the command-line and it works there, so maybe there's something different about the Nextcloud PHP environment that I'm missing.